src/pyprojroot/criterion.py

```diff
@@ -9,35 +9,50 @@
 import typing
 from os import PathLike as _PathLike
 
+
+_PathType = typing.Union[_PathLike, str]
+_CriterionType = typing.Union[
+    typing.Callable[[_PathType], bool],
+    typing.Callable[[_pathlib.Path], bool],
+    _PathType,
+    _pathlib.Path,
+    typing.Iterable[typing.Callable[[_PathType], bool]],
+    typing.Iterable[typing.Callable[[_pathlib.Path], bool]],
+]
+
 # TODO: It would be nice to have a class that encapsulates these checks,
 # so that we can implement methods like |, !, &, ^ operators
 
 # TODO: Refactor in a way that allows creation of reasons
 
 
-def as_root_criterion(criterion) -> typing.Callable:
+def as_root_criterion(
+    criterion: _CriterionType,
+) -> typing.Callable[[_pathlib.Path], bool]:
     if callable(criterion):
         return criterion
 
     # criterion must be a Collection, rather than just Iterable
     if isinstance(criterion, _PathLike):
-        criterion = [criterion]
-    criterion = list(criterion)
+        criterion_collection = [criterion]
+    else:
+
+        criterion_collection = list(criterion)  # type: ignore[arg-type]
 
     def f(path: _pathlib.Path) -> bool:
-        for c in criterion:
+        for c in criterion_collection:
             if isinstance(c, _PathLike):
                 if (path / c).exists():
                     return True
-            else:
+            if callable(c):
```

Comment:
I should revert this to `else` again. Tried to get the type inference of mypy onto the right track.

I ended up giving up here - see the ignore instruction.

Reply:
I think, ideally, this logic would all be encapsulated with classes.

Reply:
Can we leave the `else`, please? It should throw an error (ideally TypeError but that's not super important) if c isn't path-like or callable, rather than silently ignoring it.